pkg/proc/gdbserial/undo.go

```diff
@@ -158,3 +160,107 @@ func UndoIsRecording(recordingFile string) (result bool, err error) {
 
 	return bytes.Equal(marker, data), nil
 }
+
+// Fetch the output of a udbserver get_info command, split on ; and , characters.
+//
+// This is not (currently) implementing a proper parse of the data returned, just making it more
+// convenient to search.
+func undoGetInfo(conn gdbConn) []string {
+	info, err := conn.undoCmd("get_info")
+	if err != nil {
+		panic("Failed to retrieve Undo stop info.")
+	}
+	splitter := func(c rune) bool {
+		return c == ';' || c == ','
+	}
+	return strings.FieldsFunc(info, splitter)
+}
+
+func undoGetLogExtent(p *gdbProcess) (uint64, uint64) {
+	extent, err := p.conn.undoCmd("get_log_extent")
+	if err != nil {
+		panic("Failed to retrieve Undo log extent.")
+	}
+	bbcounts := strings.Split(extent, ",")
+	bbcount_min, err := strconv.ParseUint(bbcounts[0], 16, 64)
+	if err != nil {
+		panic("Failed to parse Undo log minimum.")
+	}
+	bbcount_max, err := strconv.ParseUint(bbcounts[1], 16, 64)
+	if err != nil {
+		panic("Failed to parse Undo log maximum.")
+	}
+	return bbcount_min, bbcount_max
+}
+
+// Fetch whether the replay session is currently at the end of recorded history.
+func UndoAtEndOfHistory(p *gdbProcess) bool {
+	info_fields := undoGetInfo(p.conn)
+	for _, value := range info_fields {
+		if value == "has_exited" || value == "at_event_log_end" {
+			return true
+		}
+	}
+	return false
+}
+
+// Fetch the process exit code (or zero, if not applicable) from the recording.
+func UndoGetExitCode(p *gdbProcess) int {
+	exit_code := int8(0)
+	info_fields := undoGetInfo(p.conn)
+
+	for idx, value := range info_fields {
+		if value != "has_exited" {
+			continue
+		}
+
+		// Exit status, encoded as hex, follows the has_exited string.
+		exit_status, err := strconv.ParseInt(info_fields[idx+1], 16, 16)
+		if err != nil {
+			panic("Invalid exit code.")
+		}
+
+		// Convert exit status into the form Delve usually reports - positive integer for a
+		// normal exit, negative signal number if terminated by a signal.
+		if exit_status&0x7f != 0 {
+			exit_code = -int8(exit_status & 0x7f)
+		} else {
+			exit_code = int8(exit_status >> 8)
+		}
```

Comment:
**Fixme** Status 0x7f means "stopped", not "terminated on signal 0x7f".

For maintainability I suggest making helpers with same names as the C interfaces:
```
func WTERMSIG(status int) int8 {
	return int8(status & 0x7f)
}

func WIFSIGNALED(status int) bool {
        sig := WTERMSIG(status)
	return sig != 0 && sig != 0x7f;
}

func WEXITSTATUS(status int) int8 {
	return int8((status & 0xff00) >> 8)
}
```
Then the code becomes 
```
if WIFSIGNALED(exit_status) {
        exit_code = -WTERMSIG(exit_status)
} else {
        exit_code = WEXITSTATUS(exit_status)
}
```
This is a fair bit more verbose but I think it is easier to check against the corresponding C code.

Reply:
Oooh - I've found some potentially-applicable functions in the Go `syscall` module, will investigate first whether I can just apply those.

Reply:
The `syscall.WaitStatus` type did what was required here and was easy to access, so I think we're good now if you're happy with how it looks.

Reply:
Aha, I didn't know whether if Go already had a way of doing this, but since it does it's obviously better to use that.

Reply:
I was by no means sure it'd be straightforward - Go, for instance, doesn't really (willingly) expose the fork/exec model of process handling.  But it turns out a lot of this stuff is still in there somewhere - not under the C-ish names I initially searched, it turned out.